src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -47,69 +50,118 @@ void DeepChangeChecker::find_related_tables(DeepChangeChecker::RelatedTables& ou
         auto cur_table = group->get_table(key);
         REALM_ASSERT(cur_table);
         auto incoming_link_columns = cur_table->get_incoming_link_columns();
+        auto& backlinks = complete_mapping[cur_table->get_key()];
         for (auto& incoming_link : incoming_link_columns) {
             REALM_ASSERT(incoming_link.first);
             auto& links = complete_mapping[incoming_link.first];
-            links.link_columns.push_back(incoming_link.second);
-            links.connected_tables.push_back(cur_table->get_key());
+            links.forward_links.push_back(incoming_link.second);
+            links.forward_tables.push_back(cur_table->get_key());
+            cur_table->for_each_backlink_column([&](ColKey backlink_col_key) {
+                if (any_of(key_path_array.begin(), key_path_array.end(), [&](KeyPath key_path) {
+                        return any_of(key_path.begin(), key_path.end(), [&](std::pair<TableKey, ColKey> pair) {
+                            return pair.first == cur_table->get_key() && pair.second == backlink_col_key;
+                        });
+                    })) {
+                    backlinks.backlink_tables.push_back(incoming_link.first);
+                }
+                return false;
+            });
```

Comment:
This nesting is quadratic on the number of backlinks per table and it seems unnecessary. Can you either 
move this section out one nesting level to do it only once per table and not once per link, or use `for_each_backlink_column` instead of `get_incoming_link_elements` above? (check the implementation of `get_incoming_link_elements` on how to get the forward links)

Reply:
I removed it from the loop. Using the `backlink_col_key` to determine the origin table is sufficient: 02afd82a4.